### PR TITLE
feat(frontend/shared/backend): Add Friend page tm-40

### PR DESCRIPTION
**Pull request by GEOFARL**

### CODE REVIEW

File: apps/frontend/src/pages/friends/libs/components/friend/friend.tsx [hunk 1]
```diff
@@ -41,7 +41,7 @@ const Friend: React.FC<Properties> = ({ friend }: Properties) => {
 
 	return (
 		<article className={styles["card"]}>
-			<div className={styles["card-content"]}>
+			<RouterLink className={styles["card-content"]} to={`/users/${friend.id}`}>
```
Comment (tarmiel) [inline apps/frontend/src/pages/friends/libs/components/friend/friend.tsx:44]: We already have Link component. If you faced problems with type of href prop with dynamic id,  we have configure-string helper in shared that can solve this

File: apps/frontend/src/pages/user/user.tsx [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+						alt="avatar"
+						className={styles["profile-image"]}
+						src={user.avatarUrl ?? DEFAULT_USER_AVATAR}
+					/>
+				</div>
+
+				<div className={styles["user-wrapper"]}>
+					<p
+						className={styles["fullName"]}
+					>{`${user.firstName} ${user.lastName}`}</p>
```
Comment (tarmiel) [inline apps/frontend/src/pages/user/user.tsx:74]:
```suggestion
					>{user.firstName} {user.lastName}</p>
```

**Review by tarmiel — CHANGES_REQUESTED** [review]

File: apps/frontend/src/pages/friends/libs/components/friend/friend.tsx (hunk 1 above)
Comment (GEOFARL) [reply apps/frontend/src/pages/friends/libs/components/friend/friend.tsx:44]: Yeah, couldn't figure out how to make it, thx

File: apps/frontend/src/modules/user-courses/slices/user-courses.slice.ts [hunk 3]
```diff
@@ -4,22 +4,24 @@ import { DataStatus } from "~/libs/enums/enums.js";
...
 	dataStatus: ValueOf<typeof DataStatus>;
+	myCourses: CourseDto[];
+	userCourses: Record<string, CourseDto[]>;
 };
 
 const initialState: State = {
-	courses: [],
 	dataStatus: DataStatus.IDLE,
+	myCourses: [],
+	userCourses: {},
```
Comment (what1s1ove) [inline apps/frontend/src/modules/user-courses/slices/user-courses.slice.ts:18]: should be an array

File: apps/frontend/src/pages/user/libs/components/user-button/user-button.tsx [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+const UserButton: React.FC<Properties> = ({ id }: Properties) => {
+	const [isFollowing, setIsFollowing] = useState<boolean>(
+		useAppSelector((state) =>
+			state.friends.followings.some((friend) => friend.id === id),
+		),
+	);
+	const dispatch = useAppDispatch();
+
+	const handleFollow = useCallback(() => {
```
Comment (what1s1ove) [inline apps/frontend/src/pages/user/libs/components/user-button/user-button.tsx:22]: this action should be on the `users` page level. you do not need your own wrapper for the button component

File: apps/frontend/src/pages/user/user.tsx [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+			friends: [
+				...state.friends.potentialFriends,
+				...state.friends.followers,
+				...state.friends.followings,
+			],
+			isLoading: state.userCourses.dataStatus === DataStatus.PENDING,
+		};
+	});
+
+	const user = friends.find((friend) => friend.id === userId);
```
Comment (what1s1ove) [inline apps/frontend/src/pages/user/user.tsx:37]:
```suggestion
```
you can open this page directly, so the friends reducer will be empty

File: apps/frontend/src/pages/user/user.tsx [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+} from "~/libs/hooks/hooks.js";
+import { actions } from "~/modules/user-courses/user-courses.js";
+
+import { UserButton } from "./libs/components/components.js";
+import styles from "./styles.module.css";
+
+const User: React.FC = () => {
+	const dispatch = useAppDispatch();
+	const { id } = useParams();
+	const userId = +(id as string);
```
Comment (what1s1ove) [inline apps/frontend/src/pages/user/user.tsx:24]:
```suggestion
	const userId = Number(id);
```

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: apps/frontend/src/pages/user/user.tsx (hunk 5 above)
Comment (GEOFARL) [reply apps/frontend/src/pages/user/user.tsx:39]: If we don't have friends, I redireact a user to the friends page, where they will be loaded

```ts
	if (!user) {
		return <Navigate to={AppRoute.FRIENDS} />;
	}
```

File: apps/frontend/src/pages/user/user.tsx [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+			friends: [
+				...state.friends.potentialFriends,
+				...state.friends.followers,
+				...state.friends.followings,
+			],
+			isLoading: state.userCourses.dataStatus === DataStatus.PENDING,
+		};
+	});
+
+	const user = friends.find((friend) => friend.id === userId);
```
[the remark this answers is not in the document]
Comment (GEOFARL) [reply apps/frontend/src/pages/user/user.tsx:39]: But we need a user object to display its full name, avatar and be able to follow them

[the remark this answers is not in the document]
Comment (fshabanov) [reply apps/frontend/src/pages/user/user.tsx:39]: Get user information as well with that request 🤔

[the remark this answers is not in the document]
Comment (GEOFARL) [reply apps/frontend/src/pages/user/user.tsx:39]: Okay, but why would we want to do that?

[the remark this answers is not in the document]
Comment (GEOFARL) [reply apps/frontend/src/pages/user/user.tsx:39]: Like I do not understand, should I now add a new action to the users slice to load info about a specific user? How can I get user information with a request about courses?
I do load courses by id, but before that I just get user object firstly

File: apps/frontend/src/pages/user/user.tsx (hunk 5 above)
Comment (what1s1ove) [reply apps/frontend/src/pages/user/user.tsx:39]: you need to fetch user on this page. not to take it from the `friends` reducer

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: apps/frontend/src/pages/user/user.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (fshabanov) [reply apps/frontend/src/pages/user/user.tsx:39]: You use `friends.find`, what if the user directly enters the Friend screen? In that case `friends` is an empty array, they were not loaded. You can get all needed information with 1 request - get user by id and join courses

File: apps/frontend/src/pages/user/user.tsx [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+const User: React.FC = () => {
+	const dispatch = useAppDispatch();
+	const { id } = useParams();
+	const userId = Number(id);
+	const { courses, isCoursesLoading, myId, user, userNotFound } =
+		useAppSelector((state) => {
+			return {
+				courses: state.userCourses.userCourses,
+				isCoursesLoading: state.userCourses.dataStatus === DataStatus.PENDING,
+				myId: (state.auth.user as UserAuthResponseDto).id,
```
Comment (fshabanov) [inline apps/frontend/src/pages/user/user.tsx:36]:
```suggestion
				currentUserId: (state.auth.user as UserAuthResponseDto).id,
```

File: apps/frontend/src/pages/user/user.tsx [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+				user: state.users.user,
+				userNotFound: state.users.dataStatus === DataStatus.REJECTED,
+			};
+		});
+
+	const [isFollowing, setIsFollowing] = useState<boolean>(
+		useAppSelector((state) =>
+			state.friends.followings.some((friend) => friend.id === userId),
+		),
+	);
```
Comment (fshabanov) [inline apps/frontend/src/pages/user/user.tsx:46]: Again, what if the user directly enters this screen and `state.friends.followings` is an empty array?

File: apps/frontend/src/pages/user/user.tsx [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+
+	if (userNotFound || myId === userId) {
+		return <Navigate to={AppRoute.FRIENDS} />;
+	}
+
+	if (!hasUser) {
+		return <Loader color="orange" size="large" />;
+	}
+
+	const userObject = user as UserAuthResponseDto;
```
Comment (fshabanov) [inline apps/frontend/src/pages/user/user.tsx:78]:
```suggestion
```

**Review by fshabanov — CHANGES_REQUESTED** [review]

**Review by fshabanov — CHANGES_REQUESTED** [review]

File: apps/frontend/src/modules/users/users-api.ts [hunk 11]
```diff
@@ -20,6 +20,19 @@ class UserApi extends BaseHTTPApi {
 		super({ baseUrl, http, path: APIPath.USERS, storage });
 	}
 
+	public async getById(id: number): Promise<UserAuthResponseDto> {
+		const response = await this.load(
+			this.getFullEndpoint(`${UsersApiPath.ROOT}${id}`, {}),
```
Comment (Ann-Maksay) [inline apps/frontend/src/modules/users/users-api.ts:25]:
```suggestion
			this.getFullEndpoint(UsersApiPath.$ID, { id: String(id)}),
```
Isn't it better this way? As far as I understand the second argument is exactly for parameters

File: apps/frontend/src/modules/users/users-api.ts [hunk 12]
```diff
@@ -20,6 +20,21 @@ class UserApi extends BaseHTTPApi {
 		super({ baseUrl, http, path: APIPath.USERS, storage });
 	}
 
+	public async getById(id: number): Promise<UserAuthResponseDto> {
+		const response = await this.load(
+			this.getFullEndpoint(UsersApiPath.ROOT, {
```
Comment (Ann-Maksay) [inline apps/frontend/src/modules/users/users-api.ts:25]:
```suggestion
			this.getFullEndpoint(UsersApiPath.$ID, {
```
It works this way

File: apps/backend/src/modules/friends/friend.repository.ts [hunk 13]
```diff
@@ -138,6 +140,24 @@ class FriendRepository implements Repository<UserEntity> {
...
+			.leftJoin(
+				DatabaseTableName.FRIENDS,
+				`${DatabaseTableName.USERS}.id`,
+				"=",
+				`${DatabaseTableName.FRIENDS}.following_id`,
+			)
+			.where(`${DatabaseTableName.FRIENDS}.follower_id`, "=", currentUserId)
+			.where(`${DatabaseTableName.FRIENDS}.following_id`, "=", otherUserId);
+
+		return userFollowings.length > EMPTY_FRIENDS_RESULT;
```
Comment (fshabanov) [inline apps/backend/src/modules/friends/friend.repository.ts:158]:
```suggestion
		return userFollowings.length > EMPTY_ARRAY_LENGTH;
```
We have it in shared folder

File: apps/frontend/src/pages/user/user.tsx [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+		user,
+		userNotFound,
+	} = useAppSelector((state) => {
+		return {
+			courses: state.userCourses.userCourses,
+			currentUserId: (state.auth.user as UserAuthResponseDto).id,
+			isCoursesLoading: state.userCourses.dataStatus === DataStatus.PENDING,
+			isFollowing: state.friends.isFollowing,
+			user: state.users.user,
+			userNotFound: state.users.dataStatus === DataStatus.REJECTED,
```
Comment (fshabanov) [inline apps/frontend/src/pages/user/user.tsx:44]: B19

**Review by fshabanov — CHANGES_REQUESTED** [review]

File: apps/backend/src/modules/friends/friend.controller.ts [hunk 15]
```diff
@@ -390,6 +402,43 @@ class FriendController extends BaseController {
 		};
 	}
 
+	/**
+	 * @swagger
+	 * /friends/{id}/is-following:
+	 *   get:
+	 *     description: Returns a boolean indicating whether current user follows another user
```
Comment (fshabanov) [inline apps/backend/src/modules/friends/friend.controller.ts:409]: Add `tags` property (Check #245)

**Review by fshabanov — CHANGES_REQUESTED** [review]

File: apps/frontend/src/modules/users/slices/users.slice.ts [hunk 16]
```diff
@@ -1,9 +1,37 @@
...
+import { getById } from "./actions.js";
+
+type State = {
+	dataStatus: ValueOf<typeof DataStatus>;
+	user: UserAuthResponseDto | null;
+};
+
+const initialState: State = {
+	dataStatus: DataStatus.IDLE,
+	user: null,
```
Comment (what1s1ove) [inline apps/frontend/src/modules/users/slices/users.slice.ts:16]:
```suggestion
	profileUser: null,
```
specify a name

**Review by what1s1ove — CHANGES_REQUESTED** [review]

**Review by what1s1ove — APPROVED** [review]

**Pull request merged**